File: frontend/src/types/forecast.ts
export type Direction = 'bullish' | 'bearish' | 'neutral';
export type Intensity = 'weak' | 'moderate' | 'strong' | 'extreme';
export type Regime = 
  | 'risk_on' 
  | 'risk_off' 
  | 'euphoria' 
  | 'panic' 
  | 'consolidation'
  | 'macro_stress'
  | 'reflation'
  | 'disinflation'
  | 'liquidity_expansion'
  | 'liquidity_contraction';

export type Horizon = '1D' | '1W' | '1M' | '3M' | '6M' | '1Y';

export interface Forecast {
  id: string;
  symbol: string;
  horizon: Horizon;
  forecast_date: string;
  target_date: string;
  
  direction: Direction;
  intensity: Intensity;
  
  prob_bullish: number;
  prob_bearish: number;
  prob_neutral: number;
  
  expected_return: number;
  expected_return_5th: number;
  expected_return_25th: number;
  expected_return_75th: number;
  expected_return_95th: number;
  
  detected_regime: Regime;
  regime_probability: number;
  regime_stability: number;
  
  confidence_score: number;
  stability_score: number;
  uncertainty_score: number;
  signal_robustness: number;
  
  model_weights: Record<string, number>;
  factor_contributions?: Record<string, number>;
  
  scenarios?: Record<string, { expected_return: number; probability: number }>;
  stress_scenarios?: StressScenario[];
}

export interface StressScenario {
  name: string;
  probability: number;
  expected_return: number;
  worst_case: number;
}

export interface ForecastResult {
  id: string;
  forecast_id: string;
  realized_return: number | null;
  realized_direction: Direction | null;
  directional_accuracy: number | null;
  return_error: number | null;
  probabilistic_error: number | null;
  evaluated_at: string | null;
}

export interface PerformanceMetrics {
  total_forecasts: number;
  evaluated_forecasts: number;
  directional_accuracy: number;
  average_return_error: number;
  average_confidence: number;
  confidence_calibration: number;
  accuracy_by_horizon: Record<Horizon, number>;
  accuracy_by_regime: Record<Regime, number>;
  sharpe_like_score: number;
}

File: backend/app/schemas/forecast.py
"""
Pydantic Schemas for Forecast API
"""

from pydantic import BaseModel, Field
from typing import Dict, List, Optional
from datetime import datetime, date
from uuid import UUID
from enum import Enum


class DirectionEnum(str, Enum):
    BULLISH = "bullish"
    BEARISH = "bearish"
    NEUTRAL = "neutral"


class IntensityEnum(str, Enum):
    WEAK = "weak"
    MODERATE = "moderate"
    STRONG = "strong"
    EXTREME = "extreme"


class RegimeEnum(str, Enum):
    RISK_ON = "risk_on"
    RISK_OFF = "risk_off"
    EUPHORIA = "euphoria"
    PANIC = "panic"
    CONSOLIDATION = "consolidation"
    MACRO_STRESS = "macro_stress"
    REFLATION = "reflation"
    DISINFLATION = "disinflation"
    LIQUIDITY_EXPANSION = "liquidity_expansion"
    LIQUIDITY_CONTRACTION = "liquidity_contraction"


class ForecastCreate(BaseModel):
    symbol: str
    horizon: str


class BatchForecastRequest(BaseModel):
    symbols: List[str] = Field(default=["SPX", "NDX", "DJI", "RUT"])
    horizons: List[str] = Field(default=["1D", "1W", "1M", "3M", "6M", "1Y"])


class ScenarioSchema(BaseModel):
    name: str
    probability: float
    expected_return: float
    worst_case: Optional[float] = None


class ForecastResponse(BaseModel):
    id: UUID
    symbol: str
    horizon: str
    forecast_date: datetime
    target_date: date
    
    # Direction
    direction: DirectionEnum
    intensity: IntensityEnum
    
    # Probabilities
    prob_bullish: float
    prob_bearish: float
    prob_neutral: float
    
    # Expected Returns
    expected_return: float
    expected_return_5th: float
    expected_return_25th: float
    expected_return_75th: float
    expected_return_95th: float
    
    # Regime
    detected_regime: RegimeEnum
    regime_probability: float
    regime_stability: float
    
    # Confidence
    confidence_score: float
    stability_score: float
    uncertainty_score: float
    signal_robustness: float
    
    # Model Details
    model_weights: Dict[str, float]
    factor_contributions: Optional[Dict[str, float]] = None
    
    # Scenarios
    scenarios: Optional[Dict[str, Dict]] = None
    stress_scenarios: Optional[List[ScenarioSchema]] = None
    
    class Config:
        from_attributes = True


class ForecastListResponse(BaseModel):
    forecasts: List[ForecastResponse]
    total: int
    limit: int
    offset: int


class ForecastResultResponse(BaseModel):
    id: UUID
    forecast_id: UUID
    
    realized_return: Optional[float]
    realized_direction: Optional[DirectionEnum]
    realized_volatility: Optional[float]
    
    directional_accuracy: Optional[float]
    return_error: Optional[float]
    return_error_abs: Optional[float]
    probabilistic_error: Optional[float]
    
    confidence_calibration: Optional[float]
    regime_accuracy: Optional[float]
    
    evaluated_at: Optional[datetime]
    
    class Config:
        from_attributes = True


class PerformanceSummary(BaseModel):
    total_forecasts: int
    evaluated_forecasts: int
    
    directional_accuracy: float
    average_return_error: float
    average_confidence: float
    confidence_calibration: float
    
    accuracy_by_horizon: Dict[str, float]
    accuracy_by_regime: Dict[str, float]
    
    sharpe_like_score: float
    drawdown_prediction_quality: float
